src/simulate.jl

```diff
@@ -69,18 +91,80 @@ function propagate!(sim::NestedFilterSimulation, dt)
 end
 
 function run!(sim::NestedFilterSimulation; T::Int, plot_each_timestep::Bool = false)
+    entropies = Entropies(zeros(T),zeros(T),zeros(T),zeros(T),zeros(T))
+    Maps = Map(zeros(T),zeros(T),zeros(T),zeros(T),zeros(T))
+            
+    results = Results(entropies,Maps,zeros(T))
+    
     if length(sim.times) == 0
         initialize!(sim)
     end
     for i in 1:T
         begin
-            propagate!(sim)
+            time = @timed propagate!(sim)
         end
         if plot_each_timestep
             posterior_plot(sim)
         end
+        save_results!(results, sim, i, time)
+        if i == T
+            save(string(Base.parse(Int, ENV["SLURM_ARRAY_TASK_ID"]),".jld"), "entropies", results.entropies, "ISI", sim.times, "MAP", results.map)
```

Comment:
This is not going to generalize to different machines/clusters and will throw errors if I run it on my machine.
It's better to define a general interface and keep specific things for the user scripts (which are not part of the package). I would suggest something like this
```julia
struct Recording{T1, T2, T3}
    f1::T1
    f2::T2
    data::T3
end

function Recording(f1, f2, sim::NestedFilterSimulation)
    res = f1(sim)
    data = [res]
    return Recording(f1, f2, data)
end

function update!(rec::Recording, sim)
    push!(rec.data, rec.f1(sim))
    return rec
end

function save(rec::Recording)
    rec.f2(rec.data)
end
```
You can then modify `run!` with an additional optional argument, i.e.
```julia
const NoRecording = Recording(x -> nothing, x -> nothing, Nothing[])

function run!(sim::NestedFilterSimulation; T::Int, plot_each_timestep::Bool = false, recording::Recording = NoRecording)
# modify code appropriately, using `update!` and `save` functions defined above
```
The user can then instantiate `Recording` and pass it to `run!`, e.g.
```julia
sim = NestedFilterSimulation(...)
f1(sim) = ... # code that computes entropies, maps, etc., whatever you need to record
f2(a) = ... # code that writes array `a` to a specific file, or if you don't want to do anything you type f2(a) = nothing, since you can then work with rec after running the simulation

rec = Recording(f1, f2, sim)

run!(sim, ...., recording = rec)
```

Reply:
I modified simulate.jl and created record.jl accordingly